### ISSUE TITLE
Lower performance of later version (eg: 2.5 vs. 2.0)

### ISSUE THREAD

**Issue by jcohenadad** [sct-pipeline/contrast-agnostic-softseg-spinalcord#126]
![image](https://github.com/user-attachments/assets/3c7645c0-416c-49a1-9e89-6617cb8a2e9b)

Possible avenues:
- more aggressive data augmentation

**Comment by jcohenadad** [comment]
More aggressive data augmentation was already tested at https://github.com/sct-pipeline/contrast-agnostic-softseg-spinalcord/issues/119, more specifically, [these params](https://github.com/sct-pipeline/contrast-agnostic-softseg-spinalcord/commit/600ce9bab959507789a491cd03b947b53bc2ebc8), however, results were not improved. 

Maybe @NathanMolinier or @yw7 have some insights as they used very aggressive DA, maybe there are some transfos that were not explored here?

**Comment by NathanMolinier** [comment]
All the transformations done to train totalspineseg are available [here](https://github.com/neuropoly/totalspineseg/blob/88377f2d8bb60d73736496e30126c75511f830d9/totalspineseg/utils/augment.py#L290).

**Comment by naga-karthik** [comment]
Here are the categories of transformations applied in `totalspineseg` as per [these lines](https://github.com/neuropoly/totalspineseg/blob/88377f2d8bb60d73736496e30126c75511f830d9/totalspineseg/utils/augment.py#L298-L304)

```
    Augmentation includes:
    1. Contrast augmentation (Laplace, Gamma, Histogram Equalization, Log, Sqrt, Exp, Sin, Sig, Inverse)
    2. Image from segmentation augmentation
    3. Redistribute segmentation values
    4. Artifacts augmentation (Motion, Ghosting, Spike, Bias Field, Blur, Noise)
    5. Spatial augmentation (Flip, BSpline, Affine, Elastic)
    6. Anisotropy augmentation
```

In [my set of transforms](https://github.com/sct-pipeline/contrast-agnostic-softseg-spinalcord/blob/600ce9bab959507789a491cd03b947b53bc2ebc8/monai/transforms.py#L9-L35), I cover:
1. Contrast augmentation (Gamma; L29)
2. Not applicable because it is essentially SynthSeg -- works well in case of synthesizing an image with a lot of labels but not only when SC label is present (the resulting is too unrealistic)
3. _could look into it_
4. Artifacts augmentation (Bias Field, Blur, Noise; L30-32)
5. Spatial (Affine, Flip and Elastic; L21, L25, and L34)
6. LowResolution (L28)

I notice that I _indeed_ cover most of the transforms used in TotalSpineSeg (also with larger probabilities). There are 1-2 transforms that I could add (i.e. ghosting, motion, etc. which I will consider adding)

**Comment by NathanMolinier** [comment]
I believe you should try more contrasts augmentations like Laplace, Log, Sqrt, Exp, Sin, Sig, Inverse... Gamma is definitely not enough

**Comment by naga-karthik** [comment]
do you have some references that have used these transformations (i.e. log, sqrt, exp, etc.)? curious to know how these simple math operations are helping

**Comment by NathanMolinier** [comment]
I can't tell which transformation helped the most, but I know that totalspineseg works on MP2RAGE, PSIR, proton density or CT scans without being trained on these contrasts.

**Comment by NathanMolinier** [comment]
A good test would be to run the inference multiple times on the same image after applying different transformation to evaluate the impact of each transformation.

**Comment by yw7** [comment]
> Maybe @NathanMolinier or @yw7 have some insights as they used very aggressive DA, maybe there are some transfos that were not explored here?

> ![image](https://private-user-images.githubusercontent.com/2482071/390843292-3c7645c0-416c-49a1-9e89-6617cb8a2e9b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzI4NDA3MjcsIm5iZiI6MTczMjg0MDQyNywicGF0aCI6Ii8yNDgyMDcxLzM5MDg0MzI5Mi0zYzc2NDVjMC00MTZjLTQ5YTEtOWU4OS02NjE3Y2I4YTJlOWIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTEyOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDExMjlUMDAzMzQ3WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YmVlMDNhYjRiZWE2ZTlkMDE5ZGUzZDE2YTljNzk3OGIyMmFmYzk1NjYyN2I1NTFkYzNjOTBjZjJlYTI2NGU0NSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.ciEnsWwoNw4zjlAveH3QYHDAlk03zMD5djCs4nM_mA0)

It seems that the model is able to recognize the spinal cord quite well but misses some voxels at the edges. This suggests the issue might not be the lack of augmentation but rather problems with the ground truth labels or certain data augmentation transformations. For example, some transformations might have misaligned the image and segmentation by a few voxels, or certain augmentations could have artificially altered the spinal cord boundaries (e.g., enlarging or modifying the edge voxels). As a result, the model might have learned to exclude these edge voxels from the segmentation.

Typically, data augmentation helps the model generalize to detect the object (in this case, the spinal cord) even under varying contrasts not present in the training set. However, this doesn’t seem to be the issue here. It might be worth revisiting the augmentations to identify any that could have unintentionally introduced such inconsistencies.

**Comment by naga-karthik** [comment]
> suggests the issue might not be the lack of augmentation but rather problems with the ground truth labels

@yw7 I have the same feeling as well! 

> This might be some new dataset or new data augmentation that introduced in the later versions.

precisely! that's why I kept the augmentations same as in v2.0 but added a lot more datasets causing a heavy imbalance in the number of images per each contrast
